### ISSUE TITLE
MPI_Type_get_extent_x() and threads

### ISSUE THREAD

**Issue by dalcinl** [pmodels/mpich#5871]
I recently merged MPI-4 support in mpi4py. As a consequence of these changes, now I'm using large count APIs exclusively. For example, I replaced all the usages of `MPI_Type_get_extent()` in favor of `MPI_Type_get_extent_x()`.

I'm running Fedora 35 and MPICH v3.4.1 (ch3:nemesis) from system packages. I get some strange failures in my `MPI_Op` tests. FYI, my tests are calling `MPI_Type_get_extent_x()` within user-defined reduction operations, and I believe this is create some sort of reentrancy issue with some stuff related to thread support (?). I get the following traceback:

```
Assertion failed in file src/mpi/datatype/type_get_extent_x.c at line 68: 0
/usr/lib64/mpich/lib/libmpi.so.12(MPL_backtrace_show+0x39) [0x7f40e25d40a9]
/usr/lib64/mpich/lib/libmpi.so.12(MPI_Type_get_extent_x+0x278) [0x7f40e2452a78]
...
```

After inspecting the source code, I see this line:
https://github.com/pmodels/mpich/blob/3.4.x/src/mpi/datatype/type_get_extent_x.c#L68

The implementation of the (legacy?) `MPI_Type_get_extent()` routine does not have such line.
https://github.com/pmodels/mpich/blob/3.4.x/src/mpi/datatype/type_get_extent.c#L57

I'm not really familiar with all the implementation details, but perhaps the problematic line in `MPI_Type_get_extent_x()` is not really required?  If that is the case, could you please push a fix to branch 3.4.x? Anyway, I believe I'll revert my call to `MPI_Type_get_extent()` within user-defined operations, otherwise mpi4py user-defined reductions may be unusable with current MPICH installations out there.

PS: Perhaps the issue is deeper? What if users call other MPI routines dealing with threads the same way within user-defined reductions?

**Comment by dalcinl** [comment]
Clarification: My test code does not involve threads in any way. The tests call `MPI_Allreduce()`, which in turns calls user-defined reductions, and these callbacks in turn call `MPI_Type_get_extent_x()`. So it looks like the `MPI_Allreduce()` call holds the mutex, but then `MPI_Type_get_extent_x()` is called with the mutex held.

**Comment by hzhou** [comment]
Yes, I agree that the global critical section is not needed in `type_get_extent_x`. In fact, it appears that this inconsistency is still there in mpich 4.0 and the current main branch. However, in 4.0, we made sure to exit the critical section before calling user op functions, thus preventing the assertions (on recursively entering critical sections). And before 3.4, we kind of allowed recursive critical sections -- worked, but a nightmare to chase down bugs.

> PS: Perhaps the issue is deeper? What if users call other MPI routines dealing with threads the same way within user-defined reductions?

Exactly. So the correct solution is to always exit critical sections before calling user functions. This is how it works in mpich-4.0.

>  If that is the case, could you please push a fix to branch 3.4.x? 

Maybe -- if you persistently ask for it :) But even if we do, will the distribution pick it up?

**Comment by dalcinl** [comment]
So this issue I'm having will only happen with the 3.4.x releases. Good to know! I may implement some version-specific workaround then.

> But even if we do, will the distribution pick it up?

You have a point. Likely not worth the fix. Feel free to close this issue.
